pkg/diffs/diffs.go

```diff
@@ -0,0 +1,136 @@
+	}
+
+}
+
+func (d *Diffs) findChangedFiles() error {
+	// Walk through the cloned repo
+	if err := filepath.Walk(fmt.Sprintf("%s/ci-operator/jobs/", d.DecorationPath), func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			log.Printf("prevent panic by handling failure accessing a path %q: %v\n", path, err)
+			return err
+		}
+
+		if !info.IsDir() && filepath.Ext(path) == ".yaml" {
+			masterFileContents, err := ioutil.ReadFile(fmt.Sprintf("%s/%s", d.JobConfigPath, filepath.Base(path)))
+			if err != nil {
+				log.Printf("Can't read master's file: %v", err)
+			}
+			prFileContents, err := ioutil.ReadFile(path)
+			if err != nil {
+				log.Printf("Error reading file %s: %v", path, err)
+			}
+			if bytes.Compare(masterFileContents, prFileContents) != 0 {
+				d.ChangedFiles = append(d.ChangedFiles, path)
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Diffs) getPresubmitChanges() []string {
+	var preSubmitChangedFiles []string
+	if err := d.findChangedFiles(); err != nil {
+		log.Fatal("Couldn't populate changed files.")
+	}
+	for _, changedFile := range d.ChangedFiles {
+		if strings.Contains(filepath.Base(changedFile), "presubmits.yaml") {
+			preSubmitChangedFiles = append(preSubmitChangedFiles, changedFile)
+		}
+	}
+	return preSubmitChangedFiles
+}
+
+// We return a map of maps here to make the comparing more efficient.
+func getPresubmits(file string) (map[string]map[string]prowconfig.Presubmit, error) {
+	var preSubmits PreSubmits
+
+	preSubmitsMapOfMaps := make(map[string]map[string]prowconfig.Presubmit)
+	pm := make(map[string]prowconfig.Presubmit)
+
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return preSubmitsMapOfMaps, fmt.Errorf("failed to read presubmit file (%v)", err)
+	}
+
+	if err := yaml.Unmarshal(data, &preSubmits); err != nil {
+		return preSubmitsMapOfMaps, fmt.Errorf("failed to unmarshal presubmit file (%v)", err)
+	}
+
+	for repo, preSubmitList := range preSubmits.PreSubmits {
+		for _, p := range preSubmitList {
+			pm[p.Name] = p
+			preSubmitsMapOfMaps[repo] = pm
+		}
+	}
+	return preSubmitsMapOfMaps, nil
+}
+
+// GetPresubmitsToExecute returns a map of strings with a list of Presubmits to be executed.
+func (d *Diffs) GetPresubmitsToExecute() map[string][]prowconfig.Presubmit {
+	var preSubmitsToExecute []prowconfig.Presubmit
+	preSubmitsToExecuteMap := make(map[string][]prowconfig.Presubmit)
+
+	if preSubmitChangedFiles := d.getPresubmitChanges(); len(preSubmitChangedFiles) > 0 {
+		for _, preSubmitChangedFile := range preSubmitChangedFiles {
+			prPreSubmits, _ := getPresubmits(preSubmitChangedFile)
+			masterPreSubmits, _ := getPresubmits(fmt.Sprintf("%s/%s", d.JobConfigPath, filepath.Base(preSubmitChangedFile)))
+
+			for repo, jobs := range prPreSubmits {
+				preSubmitsToExecute = []prowconfig.Presubmit{}
+				for jobName, job := range jobs {
+					if !equality.Semantic.DeepEqual(masterPreSubmits[repo][jobName].Spec, job.Spec) {
```

Comment:
I think this will crash on newly added jobs. Newly added job's `jobName` won't have an appropriate master presubmit item, and we will attempt to access such item's `Spec`.

Also, are we really sure we do not want to rehearse jobs changed in non-Spec fields? Are all these fields that harmless?

Reply:
We will have to be more specific in which fields we need to execute the job. Here I only check the `.spec`.